sql: add assignment casts for INSERTs

mgartner opened the pull request:
#### sql: rename pkg/sql/sem/tree/casts.go to cast.go

Release note: None

#### tree: add OID->OID cast map

This commit adds a new map that describes valid casts from OID to OID.
It introduces three cast contexts: explicit casts, assignment casts, and
implicit casts. See the comments for CastContext and castMap for
details.

This map will enable us to properly follow Postgres's casting behavior.
Most immediately, it will allow us to support assignment casts.

Future work includes moving volatility in castMap. In the longer term,
cast functions can be moved into castMap as well.

Release note: None

#### sql: set "char" type width to 1

The `"char"` type is a special single-character type. This commit
adds a `types.QChar` with a width one. It removes the `types.MakeQChar`
function so that it is impossible to create `"char"` types with any
other width.

Release note: None

#### sql: add assignment casts for INSERTs

Casts in Postgres are performed in one of three contexts [1]:

  1. An explicit context with `CAST(x AS T)` or `x::T`.
  2. An assignment context performed implicitly during an INSERT,
     UPSERT, or UPDATE.
  3. An implicit context during the evaluation of an expression. For
     example the DATE in `'2021-01-02'::DATE < now()` will be implicitly
     cast to a TIMESTAMPTZ so that the values can be compared.

Not all casts can be performed in all contexts. Postgres's pg_cast table
lists valid casts and specifies the maximum cast context in which each
can be performed. A cast with a max context of explicit can only be
performed in an explicit context. A cast with an assignment max context
can be performed in an explicit context or an assignment context. A cast
with an implicit max context can be performed in all contexts.

Much to my personal disappointment and frustration, there are valid
casts that are not listed in Postgres's pg_cast table. These casts are
called "automatic I/O conversions" and they allow casting most types to
and from the string types: TEXT, VARCHAR, CHAR, NAME, and "char" [2].
We cannot determine these casts' maximum cast context from the pg_cast
table, so we rely on the documentation which states that conversions to
string types are assignment casts and conversions from string types are
explicit casts [3].

--

This commit implements assignment casts for INSERTs. Follow up work will
implement assignment casts for UPSERTs and UPDATEs.

A cast performed in an assignment context has slightly different
behavior than the same cast performed in an explicit context. In an
assignment context, the cast will error if the width of the value is too
large for the given type. In an explicit context, the value will be
truncated to match the width. The one exception is assignment casts to
the special "char" type which do truncate values.

To support different cast behaviors for different contexts, a new
built-in, `crdb_internal.assignment_cast` has been introduced. This
function takes two arguments: a value and a type. Because SQL
does not have first-class types, a type cannot be passed directly to the
built-in. Instead, a `NULL` cast to a type is used as a workaround,
similar to the `json_populate_record` built-in. For example, an integer
can be assignment-cast to a string with:

    crdb_internal.assignment_cast(1::INT, NULL::STRING)

The optimizer is responsible for wrapping INSERT columns with the
assignment cast built-in function. If an insert column type `T1` is not
identical to the table's corresponding target column type `T2`, the
optimizer will check if there is a valid cast from `T1` to `T2` with a
maximum context that allows an assignment cast. If there is a such a
cast, a projection will wrap the column in the assignment cast built-in
function. If there is no such cast, a user error will be produced.

Some changes to prepared statement placeholder type inference were
required in order to better match Postgres's behavior (this is a
best-effort match thus far and there are still inconsistencies). Most
notably, widths and precision are no longer inferred for the types of
placeholders. The effect of this is that assignment casts will be
correctly added by the optimizer in order to make sure that values for
placeholders are correctly coerced to the target column type during
execution of a prepared insert.

The introduction of assignment casts fixes minor bugs and addresses some
inconsistencies with Postgres's behavior. In general, INSERTS now
successfully cast values to target table column types in more cases. As
one example, inserting a string into an integer column now succeeds:

    CREATE TABLE t (i INT)
    INSERT INTO t VALUES ('1'::STRING)

Prior to this commit there was logic that mimicked assignment casts, but
it was not correct. Bugs in the implementation caused incorrect behavior
when inserting into tables with computed columns. Most notably, a
computed column expression that referenced another column `c` was
evaluated with the value of `c` before the assignment cast was
performed. This resulted in incorrect values for computed columns in
some cases.

In addition, assignment casts make the special logic for rounding
decimal values in optbuilder obsolete. The builtin function
`crdb_internal.round_decimal_values` and related logic in optbuilder
will be removed once assignment casts are implemented for UPSERTs and
UPDATEs.

Fixes #69327
Fixes #69665

[1] https://www.postgresql.org/docs/current/typeconv.html
[2] https://www.postgresql.org/docs/13/catalog-pg-cast.html#CATALOG-PG-CAST
[3] https://www.postgresql.org/docs/13/sql-createcast.html#SQL-CREATECAST-NOTES

Release note (sql change): Implicit casts performed during INSERT
statements now more closely follow Postgres's behavior. Several minor
bugs related to these types of casts have been fixed.

mgartner committed:
sql: rename pkg/sql/sem/tree/casts.go to cast.go

Release note: None

mgartner committed:
tree: add OID->OID cast map

This commit adds a new map that describes valid casts from OID to OID.
It introduces three cast contexts: explicit casts, assignment casts, and
implicit casts. See the comments for CastContext and castMap for
details.

This map will enable us to properly follow Postgres's casting behavior.
Most immediately, it will allow us to support assignment casts.

Future work includes moving volatility in castMap. In the longer term,
cast functions can be moved into castMap as well.

Release note: None

mgartner committed:
sql: add assignment casts for INSERTs

Casts in Postgres are performed in one of three contexts [1]:

  1. An explicit context with `CAST(x AS T)` or `x::T`.
  2. An assignment context performed implicitly during an INSERT,
     UPSERT, or UPDATE.
  3. An implicit context during the evaluation of an expression. For
     example the DATE in `'2021-01-02'::DATE < now()` will be implicitly
     cast to a TIMESTAMPTZ so that the values can be compared.

Not all casts can be performed in all contexts. Postgres's pg_cast table
lists valid casts and specifies the maximum cast context in which each
can be performed. A cast with a max context of explicit can only be
performed in an explicit context. A cast with an assignment max context
can be performed in an explicit context or an assignment context. A cast
with an implicit max context can be performed in all contexts.

Much to my personal disappointment and frustration, there are valid
casts that are not listed in Postgres's pg_cast table. These casts are
called "automatic I/O conversions" and they allow casting most types to
and from the string types: TEXT, VARCHAR, CHAR, NAME, and "char" [2].
We cannot determine these casts' maximum cast context from the pg_cast
table, so we rely on the documentation which states that conversions to
string types are assignment casts and conversions from string types are
explicit casts [3].

--

This commit implements assignment casts for INSERTs. Follow up work will
implement assignment casts for UPSERTs and UPDATEs.

A cast performed in an assignment context has slightly different
behavior than the same cast performed in an explicit context. In an
assignment context, the cast will error if the width of the value is too
large for the given type. In an explicit context, the value will be
truncated to match the width. The one exception is assignment casts to
the special "char" type which do truncate values.

To support different cast behaviors for different contexts, a new
built-in, `crdb_internal.assignment_cast` has been introduced. This
function takes two arguments: a value and a type. Because SQL
does not have first-class types, a type cannot be passed directly to the
built-in. Instead, a `NULL` cast to a type is used as a workaround,
similar to the `json_populate_record` built-in. For example, an integer
can be assignment-cast to a string with:

    crdb_internal.assignment_cast(1::INT, NULL::STRING)

The optimizer is responsible for wrapping INSERT columns with the
assignment cast built-in function. If an insert column type `T1` is not
identical to the table's corresponding target column type `T2`, the
optimizer will check if there is a valid cast from `T1` to `T2` with a
maximum context that allows an assignment cast. If there is a such a
cast, a projection will wrap the column in the assignment cast built-in
function. If there is no such cast, a user error will be produced.

Some changes to prepared statement placeholder type inference were
required in order to better match Postgres's behavior (this is a
best-effort match thus far and there are still inconsistencies). Most
notably, widths and precision are no longer inferred for the types of
placeholders. The effect of this is that assignment casts will be
correctly added by the optimizer in order to make sure that values for
placeholders are correctly coerced to the target column type during
execution of a prepared insert.

The introduction of assignment casts fixes minor bugs and addresses some
inconsistencies with Postgres's behavior. In general, INSERTS now
successfully cast values to target table column types in more cases. As
one example, inserting a string into an integer column now succeeds:

    CREATE TABLE t (i INT)
    INSERT INTO t VALUES ('1'::STRING)

Prior to this commit there was logic that mimicked assignment casts, but
it was not correct. Bugs in the implementation caused incorrect behavior
when inserting into tables with computed columns. Most notably, a
computed column expression that referenced another column `c` was
evaluated with the value of `c` before the assignment cast was
performed. This resulted in incorrect values for computed columns in
some cases.

In addition, assignment casts make the special logic for rounding
decimal values in optbuilder obsolete. The builtin function
`crdb_internal.round_decimal_values` and related logic in optbuilder
will be removed once assignment casts are implemented for UPSERTs and
UPDATEs.

Fixes #69327
Fixes #69665

[1] https://www.postgresql.org/docs/current/typeconv.html
[2] https://www.postgresql.org/docs/13/catalog-pg-cast.html#CATALOG-PG-CAST
[3] https://www.postgresql.org/docs/13/sql-createcast.html#SQL-CREATECAST-NOTES

Release note (sql change): Implicit casts performed during INSERT
statements now more closely follow Postgres's behavior. Several minor
bugs related to these types of casts have been fixed.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/70722)
<!-- Reviewable:end -->

yuzefovich reviewed (commented):
I've only skimmed this a bit, but it looks like a very nice work!

> Reviewed 2 of 2 files at r1, 1 of 31 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-:-MkOdHO0Aasr0-Q9_Euq:b-jza6at)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @otan, @RaduBerinde, @rytaft, and @yuzefovich)

___
*[pkg/sql/sem/tree/type_check.go, line 460 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-MkOcCQt73We1Q7EwVh_:-MkOcCQt73We1Q7EwVha:b-cgtuce) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bf1138e073f9406deff65429ee179c89fe1ecc9/pkg/sql/sem/tree/type_check.go#L460)):*
> ```Go
> 	default:
> 		// TODO(mgartner): Use OID cast map.
> 		cast := lookupCastInfo(fromFamily, toFamily, intervalStyleEnabled, dateStyleEnabled)
> ```

nit: the change to `lookupCastInfo` is currently in the second commit but probably should be in the third commit.
___
*[pkg/sql/sem/tree/cast.go, line 74 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-MkOaTr-EjbLUeGiRqW7:-MkOaTr-EjbLUeGiRqW8:bgxdl6h) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bf1138e073f9406deff65429ee179c89fe1ecc9/pkg/sql/sem/tree/cast.go#L74)):*
> ```Go
> // context. It is only used to annotate entries in castMap and to perform
> // assertions on cast entries in the init function. It has no effect on the
> // behavior
> ```

nit: missing period.


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70722)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan, @RaduBerinde, @rytaft, and @yuzefovich)

___
*[pkg/sql/sem/tree/type_check.go, line 460 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-MkOcCQt73We1Q7EwVh_:-MkOfTLn86UhrCEt6EsK:buxe0d8) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bf1138e073f9406deff65429ee179c89fe1ecc9/pkg/sql/sem/tree/type_check.go#L460)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: the change to `lookupCastInfo` is currently in the second commit but probably should be in the third commit.
</blockquote></details>

Great catch! Fixed.
___
*[pkg/sql/sem/tree/cast.go, line 74 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-MkOaTr-EjbLUeGiRqW7:-MkOfSXN0MnSa1lXpX3W:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bf1138e073f9406deff65429ee179c89fe1ecc9/pkg/sql/sem/tree/cast.go#L74)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: missing period.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

otan commented on pkg/sql/sem/tree/cast_map_gen.sh:
```diff
@@ -0,0 +1,82 @@
+      WHEN 'i' THEN 'CastContextImplicit'
+      END
+    ),
+    provolatile
```
are we doing anything with volatile? should we?

mgartner replied:
We're not doing anything with it yet. I've removed it for now, but we'll likely want to use it to make sure we volatilities correct once we move them to this map.

otan commented on pkg/sql/sem/tree/cast.go:
```diff
@@ -39,6 +41,674 @@ import (
+	_ contextOrigin = iota
+	// pgCast specifies that a cast's maximum context is based on information in
+	// Postgres's pg_cast table.
+	pgCast
```
optional but i feel like prefixing these with `contextOrigin` makes them easier to find / read.

mgartner replied:
Done.

otan commented on pkg/sql/sem/tree/cast.go:
```diff
@@ -1493,12 +1546,6 @@ func performCastWithoutPrecisionTruncation(ctx *EvalContext, d Datum, t *types.T
 			res = NewDInt(DInt(v.Unix()))
 		case *DTimestampTZ:
 			res = NewDInt(DInt(v.Unix()))
-		case *DDate:
```
you'll need to put in a release note for this.
is this something we want to do for this commit?

mgartner replied:
Good point. I'll do this in another PR.

otan commented on pkg/sql/logictest/testdata/logic_test/cast:
```diff
@@ -1,3 +1,88 @@
+# INT4, and the INT4 -> "char" cast is explicit. Our default integer type
+# is an INT8 and INT8 -> "char" is an assignment cast.
+#
+# TODO(mgartner): This should return '{'.
```
up to you: i think rather than giving the TODO for yourself for these, file an issue and use TODO(#issue_num). that way you can abandon the baby ;) (and maybe pawn it off to uhhh an unsuspecting team)

mgartner replied:
I created #70771, but I might try to get to it before you do 😉

otan reviewed (commented):
focused mainly on experience-y bit but did skim the queries-y bits. small comments.
sorry it's on github - reviewable drives me crazy with multi-updates on  multi-commits

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Impressive work!

> Reviewed 2 of 2 files at r1, 2 of 4 files at r2, 32 of 32 files at r4, 32 of 32 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-:-MkbZZhe0hGUBgzzlJZU:b-7a3o4m)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @RaduBerinde)

___
*[pkg/sql/logictest/testdata/logic_test/cast, line 34 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-MkbRBo_0CMB_Az_wL77:-MkbRBo_0CMB_Az_wL78:b7rjvrv) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4d59bb104c2c1b4305a87477a566c40512fbd67/pkg/sql/logictest/testdata/logic_test/cast#L34)):*
> ```
> INSERT INTO assn_cast(vc) VALUES ('abc')
> 
> query T
> ```

looks like this case and the one below are duplicates
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 1421 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-MkbVHIoEZuRkUUG3At-:-MkbVHIoEZuRkUUG3At0:b-wfzj8u) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4d59bb104c2c1b4305a87477a566c40512fbd67/pkg/sql/opt/optbuilder/mutation_builder.go#L1421)):*
> ```Go
> 				colName := string(mb.tab.Column(ord).ColName())
> 				err := pgerror.Newf(pgcode.DatatypeMismatch,
> 					"value type %s doesn't match type %s of column %q",
> ```

nit: I don't think this matches the Postgres error message -- did you intend to use something different? On Postgres I got "ERROR:  column "x" is of type integer but expression is of type date" for this error.
___
*[pkg/sql/sem/tree/cast_map_test.go, line 46 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-MkbPen-DNWjiXPRGSfr:-MkbPen-DNWjiXPRGSfs:b-y0np7l) ([raw file](https://github.com/cockroachdb/cockroach/blob/743d5b0920742a6f2a48ce4a349859efbe3b121d/pkg/sql/sem/tree/cast_map_test.go#L46)):*
> ```Go
> 		_, err := tree.PerformCast(&evalCtx, srcDatum, tgtType)
> 		if err != nil && pgerror.HasCandidateCode(err) && pgerror.GetPGCode(err) == pgcode.CannotCoerce {
> 			t.Errorf("cast from %s to %s failed: %s", srcType, tgtType, err)
> ```

Do we care about any other errors produced? Maybe add a comment to explain
___
*[pkg/sql/sem/tree/cast.go, line 72 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-MkbN8aADIhXbqMQoFCu:-MkbN8aB2_NRUI3kGSEf:bsh1q7o) ([raw file](https://github.com/cockroachdb/cockroach/blob/743d5b0920742a6f2a48ce4a349859efbe3b121d/pkg/sql/sem/tree/cast.go#L72)):*
> ```Go
> 
> // contextOrigin indicates the source of information for a cast's maximum
> // context. It is only used to annotate entries in castMap and to perform
> ```

What is the "maximum" context? Is this according to the order above where explicit < assignment < implicit? Could a cast have multiple contexts?

Edit: or you could just reference the `maxContext` description below 
___
*[pkg/sql/sem/tree/cast.go, line 1241 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-MkbYN41ApFlFORlb8rl:-MkbYN420GWb5O60pYV7:b-3k2g3f) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4d59bb104c2c1b4305a87477a566c40512fbd67/pkg/sql/sem/tree/cast.go#L1241)):*
> ```Go
> // function will error if the given value is too wide for the given type. Width
> // truncation should be performed before this function for explicit casts and
> // parsing, not during assignment casts. The one exception is assignment casts
> ```

nit: this sentence is a bit confusing. It implies that explicit casts happen before assignment casts, but I think you're just trying to distinguish between the two types, right?
___
*[pkg/sql/sem/tree/cast.go, line 1435 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-Mkb_cDZ1_F8Y3fKPXLO:-Mkb_cD_5xB8KPiak42r:b-2at6z4) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4d59bb104c2c1b4305a87477a566c40512fbd67/pkg/sql/sem/tree/cast.go#L1435)):*
> ```Go
> // check on whether the datum fits the type. If truncateWidth is true, widths
> // are truncated to match the target type t. If truncateWidth is false, the
> // input datum is not truncated.
> ```

what happens if the datum doesn't fit? 

Also, seems like this only applies to some types, right? Looks like values are still being rounded etc. for numeric types.

It's a bit weird that this function is called "withoutPrecisionTruncation", but you can optionally truncate the width. Maybe add a sentence or two to explain the difference? (maybe this is what is confusing me...)


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70722)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @otan, @RaduBerinde, and @rytaft)

___
*[pkg/sql/logictest/testdata/logic_test/cast, line 34 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-MkbRBo_0CMB_Az_wL77:-MkcDoKv32gfxRFj0nf2:b2svpqh) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4d59bb104c2c1b4305a87477a566c40512fbd67/pkg/sql/logictest/testdata/logic_test/cast#L34)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

looks like this case and the one below are duplicates
</blockquote></details>

Good catch. I meant these to insert into `vc`. Fixed.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 1421 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-MkbVHIoEZuRkUUG3At-:-MkcE-RhEZJhqXe-Qs88:b-v3bliq) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4d59bb104c2c1b4305a87477a566c40512fbd67/pkg/sql/opt/optbuilder/mutation_builder.go#L1421)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: I don't think this matches the Postgres error message -- did you intend to use something different? On Postgres I got "ERROR:  column "x" is of type integer but expression is of type date" for this error.
</blockquote></details>

My intention was to leave these error messages as-is in this commit. I can follow up in a future commit to make the error message match Postgres's.
___
*[pkg/sql/sem/tree/cast_map_test.go, line 46 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-MkbPen-DNWjiXPRGSfr:-MkcHa4Q4JrmEUT7Booz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/743d5b0920742a6f2a48ce4a349859efbe3b121d/pkg/sql/sem/tree/cast_map_test.go#L46)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Do we care about any other errors produced? Maybe add a comment to explain
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/cast.go, line 72 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-MkbN8aADIhXbqMQoFCu:-MkcFAiMD80q0RyLsdw_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/743d5b0920742a6f2a48ce4a349859efbe3b121d/pkg/sql/sem/tree/cast.go#L72)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

What is the "maximum" context? Is this according to the order above where explicit < assignment < implicit? Could a cast have multiple contexts?

Edit: or you could just reference the `maxContext` description below 
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/cast.go, line 1241 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-MkbYN41ApFlFORlb8rl:-MkcGNWDBgH7H1PgmzcT:b4ziecj) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4d59bb104c2c1b4305a87477a566c40512fbd67/pkg/sql/sem/tree/cast.go#L1241)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: this sentence is a bit confusing. It implies that explicit casts happen before assignment casts, but I think you're just trying to distinguish between the two types, right?
</blockquote></details>

Good catch. I've rewritten the paragraph and hopefully it reads more clearly.
___
*[pkg/sql/sem/tree/cast.go, line 1435 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-Mkb_cDZ1_F8Y3fKPXLO:-MkcJVsTAmVf82pK9XGD:b-wpdycl) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4d59bb104c2c1b4305a87477a566c40512fbd67/pkg/sql/sem/tree/cast.go#L1435)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

what happens if the datum doesn't fit? 

Also, seems like this only applies to some types, right? Looks like values are still being rounded etc. for numeric types.

It's a bit weird that this function is called "withoutPrecisionTruncation", but you can optionally truncate the width. Maybe add a sentence or two to explain the difference? (maybe this is what is confusing me...)
</blockquote></details>

I've updated the comment to make things a little more clear. This is a code stink that we're hoping to clean up in the future. The combination of performCastWithoutPrecisionTruncation and AdjustValueToType is confusing. It's hacked together to work, but ideally the casting logic for each source and target type can live in the castMap. What makes things a bit worse is that PG's assignment cast truncation behavior is not documented anywhere as far as I can tell. So this logic is based on a few experimental observations rather than a unified fundamental theory.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
> Reviewed 3 of 33 files at r6, 2 of 2 files at r7, 32 of 32 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70722)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @otan, and @RaduBerinde)

___
*[pkg/sql/sem/tree/cast.go, line 1244 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-MkcYLdCEX0rvAP2m5lY:-MkcYLdD4_x0-J6vkTV7:bl1cddi) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e2a7ecef505e2cf713c579cdaed59529bf6f84b/pkg/sql/sem/tree/cast.go#L1244)):*
> ```Go
> // called so that an error is not returned. For assignment casts, inVal should
> // not be truncated before this function is called, so that an error is
> // returned. The one exception is for assignment casts to the special "char"
> ```

nit: "The one exception is for assignment casts to the special "char" type" -> maybe "The one exception for assignment casts is for the special "char" type"?
___
*[pkg/sql/sem/tree/cast.go, line 1245 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-MkcYhdcEQGvtbujFbaH:-MkcYhdcEQGvtbujFbaI:bumutfy) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e2a7ecef505e2cf713c579cdaed59529bf6f84b/pkg/sql/sem/tree/cast.go#L1245)):*
> ```Go
> // not be truncated before this function is called, so that an error is
> // returned. The one exception is for assignment casts to the special "char"
> // type, which truncate and do not error if the width of the datum is wider than
> ```

nit: truncate and do not error -> truncates and does not error


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
Some tests revealed that I completely forgot about casts from NULL values. Please take another look. :pray: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-:-Mkg_vYZ7XUg6O8L-tyi:b-jmllci)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @otan, @RaduBerinde, and @rytaft)

___
*[pkg/sql/sem/tree/cast.go, line 1244 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-MkcYLdCEX0rvAP2m5lY:-Mkg_al05KhwXEBssj8w:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e2a7ecef505e2cf713c579cdaed59529bf6f84b/pkg/sql/sem/tree/cast.go#L1244)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: "The one exception is for assignment casts to the special "char" type" -> maybe "The one exception for assignment casts is for the special "char" type"?
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/cast.go, line 1245 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-MkcYhdcEQGvtbujFbaH:-Mkg_q_g6jOXsj_7bLNO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e2a7ecef505e2cf713c579cdaed59529bf6f84b/pkg/sql/sem/tree/cast.go#L1245)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: truncate and do not error -> truncates and does not error
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 32 of 32 files at r9, 34 of 34 files at r10, 1 of 1 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-:-MkiB_ro9tmbVEFsgxtp:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner, @otan, and @RaduBerinde)

___
*[pkg/sql/sem/tree/cast.go, line 1811 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-MkiBS6MCqltke14EtPI:-MkiBS6MCqltke14EtPJ:b6k0mww) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf0b40957359007fc12d1e30aafb78a0ce7d0086/pkg/sql/sem/tree/cast.go#L1811)):*
> ```Go
> 				return nil, errors.Wrapf(err, "type %s", t.SQLString())
> 			}
> 			return &dd, err
> ```

nit: now you can return nil here instead of err


<!-- Sent from Reviewable.io -->

mgartner commented:
There's a problem with this implementation and prepared statements. When the expression is built during `PREPARE` we assuming a placeholder decimal has the same precision and scale as the target column's, so no assignment cast expression is added. However, during `EXECUTE` the placeholder value could have a different scale which is never adjusted because there is no assignment cast expression. 

One solution would be to add an assignment cast around every insert input column, then remove them during normalization if it can be proven they are not needed. However, this would require moving the check that the assignment cast is valid into the evaluation of the cast rather than during opt-build - we can't determine that the assignment cast in invalid if we don't know the type of the placeholder. As a result, the error you get when trying to insert an incompatibly-typed value into a column would not occur until execution-time.

rytaft commented:
> One solution would be...

Sounds like a reasonable solution to me.





<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
Very nice work!

I noticed the cases where a new `stable` shows up. Are there any cases where the an assignment cast is more volatile than the regular cast between those types? If not, we can use the existing information we have about cast volatility.

Regarding placeholders - we infer a certain type for placeholders. If the placeholder have type hints that are different, we should be able to insert assignment casts as needed during opt-build. Beyond that, any discrepancy between the expected placeholder type and what we get over the wire should be resolved by `execBind -> DecodeDatum()` - we probably just want to run assignment casts in there. (note that there are also a couple of other code paths which fill in placeholders, one for Execute and one for internal queries).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-:-MlC8k201jC8OwDVxXb_:b-wc8mi8)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @otan)



<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> I noticed the cases where a new stable shows up. Are there any cases where the an assignment cast is more volatile than the regular cast between those types? If not, we can use the existing information we have about cast volatility.

There are no cases that I am aware of. The only difference in behavior is the width truncation in an explicit cast vs an error in an assignment cast, which is not context dependent or side-effecting.

Are you suggesting that I add a mechanism for builtins so that they can have argument-dependent volatility?

> Regarding placeholders - we infer a certain type for placeholders. If the placeholder have type hints that are different, we should be able to insert assignment casts as needed during opt-build. Beyond that, any discrepancy between the expected placeholder type and what we get over the wire should be resolved by execBind -> DecodeDatum() - we probably just want to run assignment casts in there. (note that there are also a couple of other code paths which fill in placeholders, one for Execute and one for internal queries).

Brilliant. Thanks for the pointers. I'll give it a shot.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-:-MlCSdMC18tKrsmQjB9Z:b7yhfyi)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @otan)



<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> Are you suggesting that I add a mechanism for builtins so that they can have argument-dependent volatility?

I was thinking of special code in the `FunctionExpr` case in `BuildSharedProps`: if the name matches the assignment cast function, then use `tree.LookupCastVolatility`.  Maybe it would even make sense to make this a separate operator (AssignmentCastExpr) or use CastExpr and add a flag (even if it ultimately gets built as a function call).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-:-MlDTN6NDBk1iJbS3GlW:b-qcygyw)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @otan)



<!-- Sent from Reviewable.io -->

mgartner committed:
sql: set "char" type width to 1

The `"char"` type is a special single-character type. This commit adds a
`types.QChar` with a width one. It removes the `types.MakeQChar`
function so that it is impossible to create `"char"` types with any
other width.

Release note: None

mgartner reviewed (commented):
> I was thinking of special code in the FunctionExpr case in BuildSharedProps: if the name matches the assignment cast function, then use tree.LookupCastVolatility.  Maybe it would even make sense to make this a separate operator (AssignmentCastExpr) or use CastExpr and add a flag (even if it ultimately gets built as a function call).

Good idea. Done.

I made a bunch of changes in an effort to get assignment casts working correctly for prepared inserts. Please take another look everyone. :pray: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-:-MlkaPT35Y5P0mczqjUh:b-efsyc6)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @otan, @rytaft, and @yuzefovich)



<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> although it would be good to also get approval from @raduberinde or @otan since I'm a bit out of my depth here

> Reviewed 37 of 37 files at r12, 4 of 4 files at r13, 4 of 6 files at r14, 44 of 44 files at r16, 47 of 48 files at r17, 2 of 48 files at r18, 47 of 47 files at r19, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-:-MllQJOi5YTLfxkGhMzX:b-bvyvgb)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @otan)

___
*[-- commits, line 26 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-MllIlgv3xBIvgvInSvM):*
nit: It remove -> It removes
___
*[pkg/sql/sem/tree/cast.go, line 791 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-MllIFbG79zgZeAkPpGR:-MllIFbG79zgZeAkPpGS:btchct9) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fff965ced81f0559bb3c4673d2406c56408e543/pkg/sql/sem/tree/cast.go#L791)):*
> ```Go
> 				if tgt != oid.T_unknown && ent.origin != contextOriginNullConversion {
> 					panic(fmt.Sprintf(
> 						"cast from %s to %s must have an origin of contextOriginNullConversion",
> ```

nit: maybe also mention T_unknown?
___
*[pkg/sql/sem/tree/cast.go, line 227 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-MllQmpp2pk5D5syr5WJ:-MllQmpq7KoJnn_1DXlm:b-moomiy) ([raw file](https://github.com/cockroachdb/cockroach/blob/72add7600f55dd0eab06dad35247f3339b82f8fc/pkg/sql/sem/tree/cast.go#L227)):*
> ```Go
> 		// Automatic I/O conversions from "char" to other types.
> 		oid.T_bit:          {maxContext: CastContextExplicit, origin: contextOriginAutomaticIOConversion},
> 		oid.T_bool:         {maxContext: CastContextExplicit, origin: contextOriginAutomaticIOConversion},
> ```

What's the reason behind the new bool entries? Is there validation for these?


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
Thanks for the review. I realize the reviewable experience after so many rebases must not be fun.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-:-Mlp94C39-vOx-KkI8NA:b-xar5bg)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @otan, @rytaft, and @yuzefovich)

___
*[-- commits, line 26 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-MllIlgv3xBIvgvInSvM):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: It remove -> It removes
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/cast.go, line 1811 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-MkiBS6MCqltke14EtPI:-Mlp1-vmCy9UhCM3NZGW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf0b40957359007fc12d1e30aafb78a0ce7d0086/pkg/sql/sem/tree/cast.go#L1811)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: now you can return nil here instead of err
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/cast.go, line 791 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-MllIFbG79zgZeAkPpGR:-Mlp1SdwANmZ-B1Vad2q:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fff965ced81f0559bb3c4673d2406c56408e543/pkg/sql/sem/tree/cast.go#L791)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: maybe also mention T_unknown?
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/cast.go, line 227 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-MllQmpp2pk5D5syr5WJ:-Mlp3vKeDVp8ZRleMEf6:b-n1pcot) ([raw file](https://github.com/cockroachdb/cockroach/blob/72add7600f55dd0eab06dad35247f3339b82f8fc/pkg/sql/sem/tree/cast.go#L227)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

What's the reason behind the new bool entries? Is there validation for these?
</blockquote></details>

I accidentally omitted them. These `contextOriginAutomaticIOConversion` entries were manually added by me because they are not in `pg_cast`, so this was human error. I've added validation to ensure that there are entries from string types to all types, and from all types to string types.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
> Reviewed 54 of 54 files at r20, 9 of 9 files at r21, 49 of 49 files at r22, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70722)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @otan)

___
*[pkg/sql/sem/tree/cast.go, line 791 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-MllIFbG79zgZeAkPpGR:-MlpBbeE4vBKlf_v7xrw:b-csw4si) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fff965ced81f0559bb3c4673d2406c56408e543/pkg/sql/sem/tree/cast.go#L791)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

but if you get into this block `tgt` is not `T_unknown`


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70722)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @otan and @rytaft)

___
*[pkg/sql/sem/tree/cast.go, line 791 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-MllIFbG79zgZeAkPpGR:-MlpMzc_DugwhdviqfI6:b-dkd6u8) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fff965ced81f0559bb3c4673d2406c56408e543/pkg/sql/sem/tree/cast.go#L791)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

but if you get into this block `tgt` is not `T_unknown`
</blockquote></details>

Oops, good catch. Fixed.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
Fantastic job! <img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-:-MlqE0td8AgEinlXQK2t:byqhehs)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @adityamaru, @mgartner, @otan, and @rytaft)

___
*[pkg/sql/logictest/testdata/logic_test/cast, line 53 at r30](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-MlqEDAABDsYPNwpD48k:-MlqEDAABDsYPNwpD48l:b-jblu7t) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd6f55eaf0db5444131eb10a7480731a0b7a92f7/pkg/sql/logictest/testdata/logic_test/cast#L53)):*
> ```
> # Length should be 0 for both rows because CHAR is a blank-padded string type.
> query I
> SELECT length(c) FROM assn_cast
> ```

nit: for these situations, you can also do `SELECT concat('''', c, '''') FROM assn_cats` to show the values with quotes.
___
*[pkg/sql/logictest/testdata/logic_test/cast, line 152 at r30](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-MlqELXL1qOXFTIkOWKl:-MlqELXL1qOXFTIkOWKm:b-sbfwil) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd6f55eaf0db5444131eb10a7480731a0b7a92f7/pkg/sql/logictest/testdata/logic_test/cast#L152)):*
> ```
> statement ok
> EXECUTE insert_d(123.45)
> ```

Are there cases we could test where the decimal doesn't "fit" and the assignment cast fails or the value is rounded?
___
*[pkg/sql/opt/optbuilder/testdata/insert, line 1282 at r30](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-MlqEop_3nBSWQXbhBvd:-MlqEop_3nBSWQXbhBve:b-vb1w09) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd6f55eaf0db5444131eb10a7480731a0b7a92f7/pkg/sql/opt/optbuilder/testdata/insert#L1282)):*
> ```
>  └── values
>       ├── columns: column3:10!null s_default:13 rowid_default:14 column1:11 column2:12
>       └── (1, CAST(NULL AS STRING), unique_rowid(), crdb_internal.assignment_cast(' ', NULL::CHAR), crdb_internal.assignment_cast('foo', NULL::"char"))
> ```

Shouldn't we be constant-folding assignment casts like this?
___
*[pkg/sql/sem/tree/type_check.go, line 1556 at r30](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-MlqEyqJAthWZSQC21od:-MlqEyqJAthWZSQC21oe:b-8mf625) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd6f55eaf0db5444131eb10a7480731a0b7a92f7/pkg/sql/sem/tree/type_check.go#L1556)):*
> ```Go
> 	// Use the desired type family's canonical type for inference so that no
> 	// assumptions are made about a placeholders width, precision, or other
> 	// details. This is safe because no information can be lost by converting a
> ```

This is a bit confusing, should it read "no information can be lost by converting a value to the canonical type"?

I think this deserves a bit more context - the key info to include is that we only fully support populating placeholder values from pgwire for canonical types (there is no handling of things like width).

I would add some tests that exercise this, eg in TestTypeCheck. I assume that something like `2::DECIMAL(10,0) + $1` would become `2::DECIMAL(10,0) + $1:::DECIMAL`?


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-:-MlvUcS-0tm7XyNTQrWQ:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @otan, @RaduBerinde, @rytaft, and @yuzefovich)



<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70722)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @otan, @RaduBerinde, @rytaft, and @yuzefovich)

___
*[pkg/sql/logictest/testdata/logic_test/cast, line 53 at r30](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-MlqEDAABDsYPNwpD48k:-Mlu1tbN6s_C3hz4KgTd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd6f55eaf0db5444131eb10a7480731a0b7a92f7/pkg/sql/logictest/testdata/logic_test/cast#L53)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

nit: for these situations, you can also do `SELECT concat('''', c, '''') FROM assn_cats` to show the values with quotes.
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/cast, line 152 at r30](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-MlqELXL1qOXFTIkOWKl:-Mlty5zBDjDoGlNOPzzK:bdki43h) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd6f55eaf0db5444131eb10a7480731a0b7a92f7/pkg/sql/logictest/testdata/logic_test/cast#L152)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Are there cases we could test where the decimal doesn't "fit" and the assignment cast fails or the value is rounded?
</blockquote></details>

`EXECUTE insert_d(123.45)` is already testing that, no? The `d` column is `DECIMAL(10, 0)` and you can see that it was rounded in the `SELECT` query results below.

I added a similar non-prepared test. 
___
*[pkg/sql/opt/optbuilder/testdata/insert, line 1282 at r30](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-MlqEop_3nBSWQXbhBvd:-Mlu1Xfp37UD7Gg5hjA_:b-cvotv7) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd6f55eaf0db5444131eb10a7480731a0b7a92f7/pkg/sql/opt/optbuilder/testdata/insert#L1282)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Shouldn't we be constant-folding assignment casts like this?
</blockquote></details>

Yes, but we'll need a new rule for this. This is confusing, because theses assignment casts look like FunctionExprs here, but they are the AssignmentCastExprs. I've updated the formatting so this is clear. If it's alright with you, I'll add a FoldAssignmentCast normalization rule in a follow-up PR.
___
*[pkg/sql/sem/tree/type_check.go, line 1556 at r30](https://reviewable.io/reviews/cockroachdb/cockroach/70722#-MlqEyqJAthWZSQC21od:-Mlu29SE45nxL-C51_48:b-qf5v00) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd6f55eaf0db5444131eb10a7480731a0b7a92f7/pkg/sql/sem/tree/type_check.go#L1556)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This is a bit confusing, should it read "no information can be lost by converting a value to the canonical type"?

I think this deserves a bit more context - the key info to include is that we only fully support populating placeholder values from pgwire for canonical types (there is no handling of things like width).

I would add some tests that exercise this, eg in TestTypeCheck. I assume that something like `2::DECIMAL(10,0) + $1` would become `2::DECIMAL(10,0) + $1:::DECIMAL`?
</blockquote></details>

Updated the comment and added tests. 


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (approved):
> Reviewed 3 of 5 files at r28, 7 of 8 files at r29, 29 of 43 files at r30, 13 of 13 files at r31, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70722)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @otan and @rytaft)



<!-- Sent from Reviewable.io -->
